### ISSUE TITLE
Changing epub sdk

### ISSUE THREAD

**Issue by voipworld** [JideGuru/epub_viewer#49]
hello,
would you mind trying a sdk change to Readium, a more stable & well maintained epub sdk for android?
folio is not fixing some big issues, like randering a big TOC, and opening big epub files.
thanks,
Tayeb.

**Comment by JideGuru** [comment]
Hey,
You're right about Folioreader not fixing issues. I'll look into Readium and work on changing it.

**Comment by voipworld** [comment]
dont you need a flutter plugin for readium?

**Comment by JideGuru** [comment]
I'll use the swift and kotlin libraries to turn this one into a Flutter plugin

**Comment by voipworld** [comment]
otherwise just add it as a platform code and use platform channel to
open the epub into a dedicated activity, maybe.

**Comment by mrifni** [comment]
any updates on this ?

**Comment by JideGuru** [comment]
@mrifni The past few weeks have been pretty hectic for me. but I'm almost done with this.
Check the readium-android branch if you want contribute. Thanks

**Comment by mrifni** [comment]
link for the project ?

**Comment by JideGuru** [comment]
@mrifni It's this same repo. check the readium-android branch

**Comment by mrifni** [comment]
any updates on this ?

**Comment by dhruvkelawala** [comment]
Waiting for the update

**Comment by ANDRUXUIS** [comment]
any updates on this
is the plugin ready ?

**Comment by winterdl** [comment]
please change to readium sdk, Folioreader not update anymore!

**Comment by winterdl** [comment]
I've viewed the code of readium, I think we should connect direct from dart to readium js and readium css via webview. If we could implement these from flutter, we could have cross platform epub viewer. In Readium Kotlin and Readium Swift, they connect webview with readium js and css.

**Comment by JideGuru** [comment]
@winterdl Wow that's a great idea. I actually never thought of that. I'll start looking into that Immediately.

**Comment by winterdl** [comment]
I didn’t test but you should try this case, android or ios plugin uses r2-stream to create server and addPublication, from flutter we call startServer, topServer, addEpub, the plugin will return baseUrl after parse and serve publication. Flutter side we use webview to open the baseUrl, inject readium js and readium css. For simple we could try with folioReader js and folioReader css, all native ui now handle by flutter ui. If we have time, we could port r2-server and r2-stream to dart, and remove the native plugins

**Comment by mrifni** [comment]
you dont have to, its already done by some https://github.com/Mantano/mp_readium

**Comment by JideGuru** [comment]
Thanks @mrifni and @winterdl. This should make the work easier and faster

**Comment by winterdl** [comment]
I known the project but they will take time to fully port r2-streamer to dart, and the project dont have r2-navigator port. The r2-streamer-kotlin and r2-streamer-swift are mature and stable now, we should use theme as native plugin to hosting publication, and starting implement r2-navigator by flutter. When the porting of r2-streamer- dart done, we could remove the native plugin later.

**Comment by winterdl** [comment]
Good news, with the latest update of the mno packages last day, we could serve publication with flutter, no need to create native plugin:

```
final pubBox = await EpubParser().parse(filePath);
if (pubBox != null) {
  serverBloc.add(StartServer(
      [FetcherRequestHandler(pubBox.publication)]));
}
```

**Comment by winterdl** [comment]
The bad news is the flutter webview not as good as native webview, I've just tested the flutter webview to open a book, It was too laggy and slow, too dificult to swipe if we intergrate webview with pageview, the pageview didnt have preload, if we want to port UI to flutter, it will be long way to go, but we should try, and hope flutter engine will be better.

**Comment by JideGuru** [comment]
@winterdl that sounds bad. Can you send a link to the repo where you implemented this? Also, I already created a new reading branch(locally) where I'll start working on moving from Folioreader to readium (with mp-readium)

**Comment by voipworld** [comment]
Hello,
 is webView required for reading epub?
why dont we use a native text widget?

**Comment by winterdl** [comment]
Both Readium and FolioReader are using webview for rendering html layout and paging using css column. If your ebook only text, you could rendering layout by canvas, find text_composion for example or flutter_novel, or using html_view. 
I’ve just quick test, not created repo yet, but the test very simple I will post to gist later.
![B3CA2238-E7B3-4EAA-99C9-CAE2A8BD0368](https://user-images.githubusercontent.com/302480/134538255-389333c4-944a-45e6-aab1-15dc18728f99.png)
![DF3B2C9E-0470-40F1-9D51-FA801D7A08D8](https://user-images.githubusercontent.com/302480/134538282-cea4d5a5-eb9b-44d3-a191-917a5ab95888.png)

**Comment by winterdl** [comment]
The gist contains main function to open server, load epub from asset and create webview to open publication from created server: https://gist.github.com/winterdl/de0bac7ee7cd9bd932dc486c54ac06a5

**Comment by voipworld** [comment]
i think it would be great to migrate to readium either native or ported, better thant the native webView.
this would let the plugin be compatible also with windows, mac, linux and web.
what do you think?

**Comment by voipworld** [comment]
i think we better create a special branch for real test of the plugin
with the ported readium library, no?

**Comment by winterdl** [comment]
I've created a testing repo, flutter_readium, reading epub app using only flutter which based on readium project.
https://github.com/winterdl/flutter_readium . I dont have time to follow the project, hope everyone could create a repo to mainly development the idea. Thank you.

**Comment by voipworld** [comment]
Hello,
after trying your well writen example,
all i can see is the Open book button, but clicking it do nothing.
please note that i built the release apk, didnt run flutter run.

**Comment by winterdl** [comment]
I dont have android device, I just run debug on emulator. I've just tested with the code from repo, it runs okay.

<img width="600" alt="" src="https://user-images.githubusercontent.com/302480/135786663-c30b6079-1d11-466b-8857-9eccac53de40.png">

```
flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.3)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] VS Code (version 1.60.2)
```

**Comment by winterdl** [comment]
@voipworld I fixed the issue with release build, you should add android.permission.INTERNET to AndroidManifest.xml.

For now, we have 2 ways to changing sdk to readium
- Porting all readium projects to flutter (https://github.com/Mantano done almost the hard works), I've just demo a prototype at above repo, I though this is a promise way, we could write one time and run on ios and android.
- Write each native plugin, we have the branch https://github.com/JideGuru/epub_viewer/tree/readium-android, we need a prototype ios plugin.

**Comment by voipworld** [comment]
i think using the ported readium code is better in term of portability.
using platform methods mean it will work only for iOs & android.
using the ported readium code mean that we can even use it for web i
think, also windows, linux, mac.

**Comment by jmgeffroy** [comment]
Hi! 
I'm jumping into this discussion that I just discovered during a googling session. I am the initiator of the [Mantano "Iridium"](https://github.com/Mantano/iridium) open-source project. I see that @winterdl mentioned our Readium 2 ports, and I simply wanted to add that we have now opened the [source code of a Navigator](https://github.com/Mantano/mno_navigator_flutter). 
We also have developed a demo app based on your excellent [Flutter Ebook App](https://github.com/JideGuru/FlutterEbookApp). We do mention it on different pages. But for a stupid technical reason, the Git history of your Flutter Ebook App and the LICENSE have been squashed. We'll add it back quickly. Sorry for this mistake. I hope the misc. references to you and your work compensate this temporary situation. 
Basically, we have replaced the OPDS parsing and Epub viewer with ours in your app.
The Iridium project source code still contains a few rough or moving parts, and the installation and build instructions are pretty succinct.
Test versions are available on the [Play Store](https://play.google.com/store/apps/details?id=com.mantano.iridium.IridiumApp) and [TestFlight](https://testflight.apple.com/join/uHMfn7qy).
I am currently investigating some issues in the navigator (related to preloading), and I'll be happy to discuss if you wish so.
I don't want to hijack your thread, but just let you know that (I) we are open to discussing (ii) we clearly advertise the Iridium demo app as being based on your app and forward all deserved credits to you for it!
Cheers
Jean-Marie

**Comment by JideGuru** [comment]
Hi @jmgeffroy ,
Thanks for reaching out. I just tested the demo app and the viewer is awesome. My plan initially was to put all the parts of iridium together to make one viewer (makes it easier for people to use in their apps) just as you've [done here](https://github.com/Mantano/iridium/tree/main/app/lib/views/viewers). Would you be opened to us working together to achieve that or do you have plans of your own?

**Comment by jmgeffroy** [comment]
Hi @JideGuru glad to meet you, and congrats on your work!
I'm not sure to fully understand your question, since it seems that this is what we've done, basically. This [https://github.com/Mantano/iridium/tree/main/app](https://github.com/Mantano/iridium/tree/main/app) is your code (except that as I said we've squashed your git history by mistake and we'll reintroduce it asap).
Can you enlighten me? I feel I miss your point.
But yes we'll definitely be happy to cooperate on that (knowing that we are developing that in parallel with another app that relies on it).

**Comment by JideGuru** [comment]
@jmgeffroy i went through the code of the demo app and noticed that you had to import all the different parts of iridium(navigator, streamer, opds etc). This means that if a developer wants to add the viewer to their app they'll need to import all this packages as well then probably copy the code from the view page of the demo app which is totally good for people who wants like a custom integration. But I was thinking more of a quick package that puts all this together already where the developers just need to import this package and write like 5 lines of code to get things up and running. I hope that's clear now. If you've already build something like this you can point me to it. Thanks

**Comment by jmgeffroy** [comment]
OK understood. If you look at [openBook](https://github.com/Mantano/iridium/blob/c7b2369d6cca6838682b2d00be60accf89ae9bc6/app/lib/views/details/details.dart#L223), you just have to instantiate [EpubScreen](https://github.com/Mantano/iridium/blob/main/app/lib/views/viewers/epub_screen.dart) and pass it the file path.
So what we need is to extract this part, including EpubScreen and all related machinery in a separate package, such as EpubWidget. This was our intent, but we didn't finish it, I'm afraid ;-)

**Comment by JideGuru** [comment]
@jmgeffroy  Alright that makes sense. I'll wait for you to finish that so I can archive this repo then.
Thanks

**Comment by jmgeffroy** [comment]
@JideGuru OK perfect! I'll notify you in this thread as soon as it's done.
Cheers

**Comment by jmgeffroy** [comment]
Hi @JideGuru and all,
I have quickly assembled an [Iridium reader widget](https://github.com/Mantano/iridium-reader-widget) PoC, including an example.
It can be improved, this is just the very first iteration. As you can see in the code, integrating a viewer is as simple as:
```
EpubScreen.fromPath(filePath: widget.dirPath)
```
Cheers

**Comment by JideGuru** [comment]
Looks pretty good @jmgeffroy . When do you plan on publishing it?

**Comment by jmgeffroy** [comment]
Thx, I can publish it on Pub.dev tomorrow or on Monday. I also have created a separate Iridium-demo repo forked from your repo. It doesn't use the widget yet, but it will be done shortly.

**Comment by dymxin** [comment]
> Hi!
> I'm jumping into this discussion that I just discovered during a googling session. I am the initiator of the [Mantano "Iridium"](https://github.com/Mantano/iridium) open-source project. I see that @winterdl mentioned our Readium 2 ports, and I simply wanted to add that we have now opened the [source code of a Navigator](https://github.com/Mantano/mno_navigator_flutter).
> We also have developed a demo app based on your excellent [Flutter Ebook App](https://github.com/JideGuru/FlutterEbookApp). We do mention it on different pages. But for a stupid technical reason, the Git history of your Flutter Ebook App and the LICENSE have been squashed. We'll add it back quickly. Sorry for this mistake. I hope the misc. references to you and your work compensate this temporary situation.
> Basically, we have replaced the OPDS parsing and Epub viewer with ours in your app.
> The Iridium project source code still contains a few rough or moving parts, and the installation and build instructions are pretty succinct.
> Test versions are available on the [Play Store](https://play.google.com/store/apps/details?id=com.mantano.iridium.IridiumApp) and [TestFlight](https://testflight.apple.com/join/uHMfn7qy).
> I am currently investigating some issues in the navigator (related to preloading), and I'll be happy to discuss if you wish so.
> I don't want to hijack your thread, but just let you know that (I) we are open to discussing (ii) we clearly advertise the Iridium demo app as being based on your app and forward all deserved credits to you for it!
> Cheers
> Jean-Marie

Such a beautiful work. I tried the TestFlight, and noticed that the bookmark function doesn’t seem to be working.

**Comment by jmgeffroy** [comment]
Hi @dymxin 
> Such a beautiful work. I tried the TestFlight, and noticed that the bookmark function doesn’t seem to be working.
Thanks! Much appreciated. I'll double-check this issue and publish an update. Stay tuned!

**Comment by dymxin** [comment]
> 

Great! Also if there is a possibility to include last location functionality. Whereby on open, a book goes to the last page viewed. Much Love

**Comment by voipworld** [comment]
Hi @jmgeffroy
yes i saw the new unified repo, good step for the project.
i'm currently switching to beta to build and test it, i will test the reader widget in term of usability and accessibility for screen reader users and give fidback on the corresponding repo.
i'm pretty happy with this new component, i have been dreaming for a pure flutter epub reading widget for a long time. there was epub_view, but no arabic support.

**Comment by voipworld** [comment]
i tryed the widget. the reading is pretty easy and well compatible with screen readers.
the only issue is the buttons is not labeled, i think you're just using iconButton, can you use textButton.icon? so you can ofer labeling.

**Comment by voipworld** [comment]
Hello @@jmgeffroy
i tested the widget very well, i opened two issues on the newer unified repo.
following up!

**Comment by jmgeffroy** [comment]
Great, thanks a lot, we'll look at your issues right now.
Cheers

**Comment by JideGuru** [comment]
@jmgeffroy Great work you've done with Iridium. I think I'll go ahead and discontinue this plugin now so everyone can move to Iridium.

**Comment by ABDERRAHMANE-OUALI** [comment]
@jmgeffroy  any news regarding the the reader? did you publish it to pub.dev
